Ft receipts

tonyolori opened the pull request.

Commit:
Add receipt sharing functionality to Transaction Service

- Introduced `SignedLink` and `ReceiptDocument` entities for managing shareable links and receipt documents.
- Created `IReceiptService` interface and `ReceiptService` implementation for receipt generation and link sharing, integrating with Cloudinary.
- Added `GetTransactionReceiptQuery` and its handler to retrieve receipt links, with logic for checking existing links.
- Enhanced architecture with new DTOs, including `ReceiptDto`, for improved data transfer.
- Updated unit and integration tests to ensure reliability of the new features.

🚀 Anthony Olori

Commit:
Add receipt and signed link repositories with tests

- Introduced `IReceiptDocumentRepository` and `ISignedLinkRepository` interfaces for managing receipt documents and signed links.
- Implemented these interfaces in `ReceiptDocumentRepository` and `SignedLinkRepository` classes using Entity Framework Core for CRUD operations.
- Updated `ServiceCollectionExtensions.cs` to register the new repositories with the dependency injection container.
- Added unit tests for both repositories in `ReceiptDocumentRepositoryTests` and `SignedLinkRepositoryTests`, covering various scenarios.
- Utilized an in-memory database for tests to ensure isolation and repeatability.

🚀 Anthony Olori

Commit:
Refactor TransactionsController and add validation methods

- Replaced `CreateShareableReceipt` with `GetTransactionReceipt` for retrieving transaction receipts with input validation.
- Introduced `ValidateReceiptLink` method for validating receipt links, including error handling and logging.
- Added `ValidateReceiptLinkQueryHandler` to process link validation requests.
- Created `ValidateReceiptLinkQuery` to encapsulate validation requests.
- Expanded `TransactionsControllerTests` with new unit tests for the updated methods.
- Updated project file to include `Moq` for mocking in unit tests.

🚀 Anthony Olori

Barkty commented on src/TransactionService.Api/Controllers/TransactionsController.cs:
```diff
@@ -19,16 +19,77 @@ public async Task<IActionResult> GetTransaction([FromRoute] Guid id)
+            if (expirationHours <= 0 || expirationHours > 168)
+            {
+                return BadRequest(new { error = "expirationHours must be between 1 and 168 hours" });
+            }
```
Can you create a validation class for this and pass the variables in it? Check the validation class I did for International and domestic transaction

Barkty commented on src/TransactionService.Api/Controllers/TransactionsController.cs:
```diff
@@ -19,16 +19,77 @@ public async Task<IActionResult> GetTransaction([FromRoute] Guid id)
+            }
+        }
+
+        [HttpGet("/receipt/validate")]
+        public async Task<IActionResult> ValidateReceiptLink([FromQuery] string url)
+        {
+            if (string.IsNullOrWhiteSpace(url))
+            {
+                return BadRequest(new { error = "url parameter is required" });
+            }
+
+            try
+            {
+                var query = new ValidateReceiptLinkQuery(url);
+                var isValid = await _mediator.Send(query);
+                
+                return Ok(new { isValid, url });
+            }
+            catch (Exception ex)
+            {
```
Is this a preliminary to sharing receipts?

tonyolori replied:
more or less.

Barkty commented on src/TransactionService.Api/Controllers/TransactionsController.cs:
```diff
@@ -19,16 +19,77 @@ public async Task<IActionResult> GetTransaction([FromRoute] Guid id)
+                return BadRequest(new { error = ex.Message });
+            }
+            catch (Exception ex)
+            {
+                return StatusCode(500, new { error = "An error occurred while generating the receipt" });
+            }
+        }
+
+        [HttpGet("/receipt/validate")]
+        public async Task<IActionResult> ValidateReceiptLink([FromQuery] string url)
+        {
+            if (string.IsNullOrWhiteSpace(url))
+            {
+                return BadRequest(new { error = "url parameter is required" });
+            }
+
+            try
+            {
+                var query = new ValidateReceiptLinkQuery(url);
+                var isValid = await _mediator.Send(query);
+                
+                return Ok(new { isValid, url });
+            }
+            catch (Exception ex)
+            {
+                return StatusCode(500, new { error = "An error occurred while validating the link" });
+            }
         }
     }
 
 }
 
 
-public record CreateShareRequest(int ExpiresInSeconds);
+public record ReceiptResponse(
+    Guid TransactionId,
+    string ShareableUrl,
+    DateTime ExpiresAt,
+    string? DocumentUrl = null
+);
```
You can use the DTO instead. And we can keep things in their own files to keep things neat and if we want to expand features

Barkty commented on src/TransactionService.Application/Handlers/GetTransactionReceiptQueryHandler.cs:
```diff
@@ -0,0 +1,94 @@
+                request.TransactionId, request.RequestedBy);
+
+            // Validate transaction exists
+            var transaction = _context.Transactions
+                .FirstOrDefault(t => t.Id == request.TransactionId);
+
```
There is a transaction repository that communicates with the DbContext. So we don't have to drag that guy into every conversation

tonyolori replied:
alright, done

Barkty commented on src/TransactionService.Application/Handlers/GetTransactionReceiptQueryHandler.cs:
```diff
@@ -0,0 +1,94 @@
+                           sl.IsActive && 
+                           sl.ExpiresAt > DateTime.UtcNow)
+                .OrderByDescending(sl => sl.CreatedAt)
+                .FirstOrDefault();
```
Create a Repository for SignedLinks as well

Barkty commented on src/TransactionService.Application/Handlers/GetTransactionReceiptQueryHandler.cs:
```diff
@@ -0,0 +1,94 @@
+                _logger.LogInformation("Found existing valid receipt link for transaction {TransactionId}", request.TransactionId);
+                
+                var existingDocument = _context.ReceiptDocuments
+                    .FirstOrDefault(rd => rd.TransactionId == request.TransactionId);
```
A repository for this model as well

Barkty commented on src/TransactionService.Domain/Entities/ReceiptDocument.cs:
```diff
@@ -0,0 +1,50 @@
+        [Required]
+        public Guid TransactionId { get; set; }
+
+        [Required]
+        [MaxLength(500)]
+        public string DocumentUrl { get; set; } = string.Empty;
+
+        [Required]
+        [MaxLength(200)]
+        public string CloudinaryPublicId { get; set; } = string.Empty;
+
```
What is the difference between the two?. How are we getting the document url?

tonyolori replied:
DocumentUrl is the public link used to view or download the file from Cloudinary, while CloudinaryPublicId is the unique internal identifier used to manage the file programmatically (e.g., delete or update it). In essence, the URL is for access, and the public ID is for control.

Barkty commented on src/TransactionService.Domain/Entities/SignedLink.cs:
```diff
@@ -0,0 +1,66 @@
+        public Guid TransactionId { get; set; }
+
+        [Required]
+        [MaxLength(50)]
+        public string ResourceType { get; set; } = "Receipt";
+
```
If we are using the Resource type, it means we can have other resources that aren't transactions and won't have transaction IDs. Shouldn't the TransactionId be ResourceId?

tonyolori replied:
thats true, i was thinking from the perspective that whatever we're dealing with would always be linked to a transaction. but changing the name would be fine.

Barkty commented on src/TransactionService.Infrastructure/Data/AppDBContext.cs:
```diff
@@ -92,6 +96,74 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+                entity.Property(e => e.Id)
+                    .ValueGeneratedNever();
+
+                entity.Property(e => e.TransactionId)
+                    .IsRequired();
+
+                entity.Property(e => e.ResourceType)
+                    .HasMaxLength(50)
+                    .IsRequired();
```
The transaction ID I mentioned. What if we want users to be able to share other things?

tonyolori replied:
i see, im fine with renaming to ResourceId. is that alright?

Barkty commented on src/TransactionService.Infrastructure/Services/CloudinaryService.cs:
```diff
@@ -0,0 +1,118 @@
+            _cloudName = _configuration["Cloudinary:CloudName"] ?? "drfy6rjlw";
+            _apiKey = _configuration["Cloudinary:ApiKey"] ?? "968343478423413";
+            _apiSecret = _configuration["Cloudinary:ApiSecret"] ?? "VXVgmx4InQK9kRC3VCsmSaNxb4k";
+        }
```
We won't be passing env values in code

tonyolori replied:
agreed, resolved.

Barkty commented on tests/TransactionService.Domain.Tests/Entities/ReceiptDocumentTests.cs:
```diff
@@ -0,0 +1,104 @@
+            receiptDocument.TransactionId.Should().BeEmpty();
+            receiptDocument.DocumentUrl.Should().Be(string.Empty);
+            receiptDocument.CloudinaryPublicId.Should().Be(string.Empty);
+            receiptDocument.CreatedAt.Should().BeCloseTo(DateTime.UtcNow, TimeSpan.FromSeconds(1));
```
Why not use the Assert?

Barkty commented on tests/TransactionService.Domain.Tests/Entities/ReceiptDocumentTests.cs:
```diff
@@ -0,0 +1,104 @@
+            var action = () => new ReceiptDocument(Guid.Empty, documentUrl, cloudinaryPublicId);
+            action.Should().Throw<ArgumentException>()
+                .WithMessage("TransactionId cannot be empty*");
+        }
```
Same here

Barkty commented on tests/TransactionService.Domain.Tests/Entities/ReceiptDocumentTests.cs:
```diff
@@ -0,0 +1,104 @@
+            var action = () => new ReceiptDocument(transactionId, documentUrl, cloudinaryPublicId);
+            action.Should().Throw<ArgumentException>()
+                .WithMessage("DocumentUrl cannot be empty*");
+        }
```
Same here too

Barkty commented on tests/TransactionService.Domain.Tests/Entities/SignedLinkTests.cs:
```diff
@@ -0,0 +1,148 @@
+            signedLink.ExpiresAt.Should().Be(expiresAt);
+            signedLink.ResourceType.Should().Be("Receipt");
+            signedLink.IsActive.Should().BeTrue();
+            signedLink.CreatedAt.Should().BeCloseTo(DateTime.UtcNow, TimeSpan.FromSeconds(1));
```
Assert here too

Barkty reviewed (changes requested):
Nice work here. Just a few more steps to go. Use the Assert keyword in your test so other team members can easily understand what your test intentions are

Commit:
Merge branch 'main' of https://github.com/Dev-Mentorship-Programme/transaction-service into ft-receipts

Commit:
Add receipt sharing feature with validation and tests

- Updated `TransactionsController` to include `IReceiptRequestValidator` for parameter validation.
- Added new entities `ReceiptDocument` and `SignedLink` to the database schema with migration.
- Enhanced `CloudinaryService` to require configuration values.
- Implemented `ReceiptRequestValidator` for centralized validation logic.
- Added unit tests for `ReceiptRequestValidator` and updated controller tests.

🚀 Anthony Olori

Commit:
update tests to remove fluent assertions

Commit:
Refactor transaction handling with repository pattern

- Replaced direct database context access with repository interfaces (`ITransactionRepository`, `ISignedLinkRepository`, `IReceiptDocumentRepository`).
- Updated `GetTransactionReceiptQueryHandler` to use new repository methods for data retrieval.
- Modified `SignedLink` entity to use `ResourceId` instead of `TransactionId`.
- Adjusted `PdfDocumentGenerator` to align with the new data model.
- Updated tests to reflect changes in repositories and data model.

🚀 Anthony Olori
